freqtrade/freqai/data_drawer.py

```diff
@@ -0,0 +1,642 @@
+        )
+        self.historic_predictions_path = Path(self.full_path / "historic_predictions.pkl")
+        self.pair_dictionary_path = Path(self.full_path / "pair_dictionary.json")
+        self.follow_mode = follow_mode
+        if follow_mode:
+            self.create_follower_dict()
+        self.load_drawer_from_disk()
+        self.load_historic_predictions_from_disk()
+        self.training_queue: Dict[str, int] = {}
+        self.history_lock = threading.Lock()
+        self.old_DBSCAN_eps: Dict[str, float] = {}
+
+    def load_drawer_from_disk(self):
+        """
+        Locate and load a previously saved data drawer full of all pair model metadata in
+        present model folder.
+        :return: bool - whether or not the drawer was located
+        """
+        exists = self.pair_dictionary_path.is_file()
+        if exists:
+            with open(self.pair_dictionary_path, "r") as fp:
+                self.pair_dict = json.load(fp)
+        elif not self.follow_mode:
+            logger.info("Could not find existing datadrawer, starting from scratch")
+        else:
+            logger.warning(
+                f"Follower could not find pair_dictionary at {self.full_path} "
+                "sending null values back to strategy"
+            )
+
+        return exists
+
+    def load_historic_predictions_from_disk(self):
+        """
+        Locate and load a previously saved historic predictions.
+        :return: bool - whether or not the drawer was located
+        """
+        exists = self.historic_predictions_path.is_file()
+        if exists:
+            with open(self.historic_predictions_path, "rb") as fp:
+                self.historic_predictions = cloudpickle.load(fp)
+            logger.info(
+                f"Found existing historic predictions at {self.full_path}, but beware "
+                "that statistics may be inaccurate if the bot has been offline for "
+                "an extended period of time."
+            )
+        elif not self.follow_mode:
+            logger.info("Could not find existing historic_predictions, starting from scratch")
+        else:
+            logger.warning(
+                f"Follower could not find historic predictions at {self.full_path} "
+                "sending null values back to strategy"
+            )
+
+        return exists
+
+    def save_historic_predictions_to_disk(self):
+        """
+        Save data drawer full of all pair model metadata in present model folder.
+        """
+        with open(self.historic_predictions_path, "wb") as fp:
+            cloudpickle.dump(self.historic_predictions, fp, protocol=cloudpickle.DEFAULT_PROTOCOL)
+
+    def save_drawer_to_disk(self):
+        """
+        Save data drawer full of all pair model metadata in present model folder.
+        """
+        with open(self.pair_dictionary_path, "w") as fp:
+            json.dump(self.pair_dict, fp, default=self.np_encoder)
```

Comment:
I assume this is running freqently / has a lot of data passing through?
In which case, i'd suggest using rapidjson.dump for it - it's a LOT faster ...